### PR TITLE
Add type for Node.js process.env

### LINKED ISSUE #11615: Type of process.env should be { [key: string]: string }
This will allow to typescript to trigger compile error :

``` js
    const x: number = process.env.x;
```

See Microsoft/TypeScript/issues/11239

### PR DESCRIPTION
**Pull request by xeoneux**
Fixes #11615 & #13633

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

### CODE REVIEW

**Comment by mhegazy** [comment]
CI tests are failing. can you please address the failures.

**Comment by xeoneux** [comment]
@mhegazy I do not get why they are failing. Could really use a hand!

**Comment by mhegazy** [comment]
`env` was `any` before, so tests in the packages that are failing now were using with no working. changing the definition made it an error  (Pre TS 2.2) if you access a property on `env`.

You could just update the tests to use element access instead of property access, i.e:
```ts
process.env["PASSPORT_TWITTER_CONSUMER_KEY"]
```
instead of 
```ts
process.env.PASSPORT_TWITTER_CONSUMER_KEY
```

**Comment by xeoneux** [comment]
Thanks @mhegazy 
So these tests are passing in TS 2.2+ but failing pre-TS 2.2? Why's that?
Will I have to change the tests for every definition that depends on `process.env`?

**Comment by mhegazy** [comment]
> So these tests are passing in TS 2.2+ but failing pre-TS 2.2? Why's that?

https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#dotted-property-for-types-with-string-index-signatures

> Will I have to change the tests for every definition that depends on process.env?

yes. we need all tests to be in a consistent state before we can merge any changes.

File: types/dockerode/dockerode-tests.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@ const docker4 = new Docker({ host: '127.0.0.1', port: 3000 });
 
 const docker5 = new Docker({
   host: '192.168.1.10',
-  port: process.env.DOCKER_PORT || 2375,
+  port: parseInt(process.env["DOCKER_PORT"]) || 2375,
```
Comment (mhegazy) [inline types/dockerode/dockerode-tests.ts]: missing a radex parameter. consider instead `+process.env["DOCKER_PORT"]`

**Comment by mhegazy** [comment]
Two more errors in dockerode.

**Comment by mhegazy** [comment]
@andy-ms @kadler15 @RyanCavanaugh @blakeembrey @guybedford @Loghorn can you please review this change.

**Comment by guybedford** [comment]
Sure, does the PR mean that all uses in code of `process.env.x` would need to be `process.env['x']`? I'm not sure that is something that should be imposed on user code?

What's the goal here? To get the PATH property available? Is there not another way?

**Comment by mhegazy** [comment]
> Sure, does the PR mean that all uses in code of process.env.x would need to be process.env['x']? I'm not sure that is something that should be imposed on user code?

Yes for users with TS prior to 2.2.

File: types/node/index.d.ts [hunk 2]
```diff
@@ -342,6 +342,11 @@ declare namespace NodeJS {
         system: number;
     }
 
+    export interface Env {
+        PATH: string;
```
Comment (andy-ms) [inline types/node/index.d.ts]: Why specify this? It should work anyway because of the string indexer.

File: types/node/index.d.ts [hunk 3]
```diff
@@ -342,6 +342,11 @@ declare namespace NodeJS {
...
+        [key: string]: string;
```
Comment (andy-ms) [inline types/node/index.d.ts]: Should be `: string | undefined;`.

**Comment by andy-ms** [comment]
I think this should wait for the node version 8 release since it will cause lots of compiler errors for anyone on ts<2.2.

**Comment by blakeembrey** [comment]
This is a useful change and something I've supported on @types for a long time. However, part of the change is incorrect - the `Env` type should only be used for `process.env` and not in the other input positions as that will require users to specify every required key for things like `fork()`. It'd be better to make those other positions `object`.

**Comment by mhegazy** [comment]
Thanks for your contribution. This PR has failing CI tests and can not be merged in at the time being. For housekeeping purposes we are closing stale PRs. If you'd still like to merge this code in, please open a new PR that has been merged and rebased with the `master` branch.

**Comment by zixia** [comment]
We should merge this PR because it's really useful!

Today I ran into an issue that I write by mistake like: 

```ts
!!process.env('WECHATY_DOCKER')
```

vscode did not warn me and this also passed the tslint!

**Pull request closed without merging**